File: crossfit-timer-web/src/utils/timerEngine.js
/**
 * Timer Engine - Date-based timer calculations to prevent drift
 * Matches the iOS TimerEngine.swift implementation pattern
 */

/**
 * Calculate time remaining until target end time
 * @param {number} targetEndTime - Target end timestamp (Date.now() format)
 * @returns {number} Seconds remaining (can be 0 if ended)
 */
export const calculateTimeRemaining = (targetEndTime) => {
  const now = Date.now();
  const remaining = Math.max(0, targetEndTime - now);
  return remaining / 1000; // Convert to seconds
};

/**
 * Calculate progress percentage for an interval
 * @param {number} targetEndTime - Target end timestamp
 * @param {number} duration - Total duration in seconds
 * @returns {number} Progress from 0 to 1
 */
export const calculateProgress = (targetEndTime, duration) => {
  const timeRemaining = calculateTimeRemaining(targetEndTime);
  const elapsed = duration - timeRemaining;
  return Math.min(1, Math.max(0, elapsed / duration));
};

/**
 * Check if interval has ended
 * @param {number} targetEndTime - Target end timestamp
 * @returns {boolean} True if interval has ended
 */
export const hasIntervalEnded = (targetEndTime) => {
  return Date.now() >= targetEndTime;
};

/**
 * Create a new target end time for an interval
 * @param {number} durationSeconds - Duration in seconds
 * @returns {number} Target end timestamp
 */
export const createTargetEndTime = (durationSeconds) => {
  return Date.now() + (durationSeconds * 1000);
};

/**
 * Calculate remaining time when paused
 * @param {number} targetEndTime - Original target end time
 * @returns {number} Seconds remaining at pause time
 */
export const calculatePausedRemaining = (targetEndTime) => {
  return calculateTimeRemaining(targetEndTime);
};

/**
 * Create new target end time when resuming from pause
 * @param {number} pausedRemaining - Seconds remaining when paused
 * @returns {number} New target end timestamp
 */
export const createResumeEndTime = (pausedRemaining) => {
  return createTargetEndTime(pausedRemaining);
};

File: Timer_App/CircuitTimer/Services/TimerEngine.swift
//
//  TimerEngine.swift
//  CircuitTimer
//
//  Core date-based timer logic that survives iOS background suspension
//

import Foundation

/// Timer engine using date-based calculations for background safety
///
/// This class implements a date-based timer that remains accurate even when
/// the app is suspended by iOS. Instead of counting down, it stores a target
/// end date and calculates remaining time on demand.
class TimerEngine {
    // MARK: - Properties

    /// Target end date for the current interval (nil when not running)
    private(set) var intervalEndDate: Date?

    /// Time remaining when paused (in seconds)
    private var pausedTimeRemaining: TimeInterval = 0

    /// Whether the timer is currently running
    var isRunning: Bool {
        return intervalEndDate != nil
    }

    /// Whether the timer is currently paused
    var isPaused: Bool {
        return !isRunning && pausedTimeRemaining > 0
    }

    // MARK: - Public Methods

    /// Start a new interval with the specified duration
    /// - Parameter duration: Interval duration in seconds
    func startInterval(duration: TimeInterval) {
        guard duration > 0 else { return }
        intervalEndDate = Date().addingTimeInterval(duration)
        pausedTimeRemaining = 0
    }

    /// Get the time remaining in the current interval
    /// - Returns: Time remaining in seconds (0 if not running)
    func timeRemaining() -> TimeInterval {
        if let endDate = intervalEndDate {
            let remaining = endDate.timeIntervalSinceNow
            return max(0, remaining)
        } else if pausedTimeRemaining > 0 {
            return pausedTimeRemaining
        }
        return 0
    }

    /// Check if the current interval has ended
    /// - Returns: True if the interval end date has passed
    func hasIntervalEnded() -> Bool {
        guard let endDate = intervalEndDate else { return false }
        return Date() >= endDate
    }

    /// Pause the timer, storing remaining time
    func pause() {
        guard let endDate = intervalEndDate else { return }
        pausedTimeRemaining = max(0, endDate.timeIntervalSinceNow)
        intervalEndDate = nil
    }

    /// Resume the timer from paused state
    func resume() {
        guard pausedTimeRemaining > 0 else { return }
        intervalEndDate = Date().addingTimeInterval(pausedTimeRemaining)
        pausedTimeRemaining = 0
    }

    /// Reset the timer to initial state
    func reset() {
        intervalEndDate = nil
        pausedTimeRemaining = 0
    }

    /// Get the progress percentage of the current interval
    /// - Parameter totalDuration: Total duration of the interval
    /// - Returns: Progress from 0.0 (start) to 1.0 (end)
    func progress(totalDuration: TimeInterval) -> Double {
        guard totalDuration > 0 else { return 0 }
        let remaining = timeRemaining()
        let elapsed = totalDuration - remaining
        return min(1.0, max(0.0, elapsed / totalDuration))
    }

    /// Calculate how much time has passed since interval started
    /// - Parameter totalDuration: Total duration of the interval
    /// - Returns: Elapsed time in seconds
    func elapsedTime(totalDuration: TimeInterval) -> TimeInterval {
        let remaining = timeRemaining()
        return totalDuration - remaining
    }
}

// MARK: - Debug Description
extension TimerEngine: CustomDebugStringConvertible {
    var debugDescription: String {
        if let endDate = intervalEndDate {
            return "TimerEngine(endDate: \(endDate), remaining: \(timeRemaining())s)"
        } else if pausedTimeRemaining > 0 {
            return "TimerEngine(paused: \(pausedTimeRemaining)s)"
        } else {
            return "TimerEngine(idle)"
        }
    }
}
